cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,858 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+ * @param[in]  grid_dim      X/Y dimensions of the grid to be initialized
+ * @param[in]  tg_dim        X/Y dimensions of tile groups in grid
+ * @param[in]  name          Kernel name to be executed on tile groups in grid
+ * @param[in]  argc          Number of input arguments to kernel
+ * @param[in]  argv          List of input arguments to kernel
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+int hb_mc_grid_init (device_t *device, hb_mc_dimension_t grid_dim, hb_mc_dimension_t tg_dim, char* name, uint32_t argc, uint32_t argv[]) {
+	int error; 
+	for (hb_mc_idx_t tg_id_x = 0; tg_id_x < grid_dim.x; tg_id_x ++) { 
+		for (hb_mc_idx_t tg_id_y = 0; tg_id_y < grid_dim.y; tg_id_y ++) { 
+			hb_mc_coordinate_t tg_id = hb_mc_coordinate(tg_id_x, tg_id_y);
+			error = hb_mc_tile_group_enqueue(device, device->num_grids, tg_id, grid_dim, tg_dim, name, argc, argv); 
+			if (error != HB_MC_SUCCESS) { 
+				bsg_pr_err("%s: failed to initialize tile group (%d,%d) of grid %d.\n", __func__, tg_id_x, tg_id_y, device->num_grids);
+				return HB_MC_FAIL;
+			}
+		}
+	}
+	device->num_grids ++;
+	return HB_MC_SUCCESS;
 }
 
-/*
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor, which must be obtained from hb_mc_init_host.
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_init (uint8_t *fd, eva_id_t eva_id, char *elf, tile_t *tiles, uint32_t num_tiles) {
-	
-	int error = hb_mc_fifo_init(fd); 
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_init() --> hb_mc_init_host(): failed to initialize host.\n");
+int hb_mc_tile_group_allocate_tiles (device_t *device, tile_group_t *tg){
+	int error;
+	if (tg->dim.x > device->mesh->dim.x){
+		bsg_pr_err("%s: tile group X dimension (%d) larger than mesh X dimension (%d).\n", __func__, tg->dim.x, device->mesh->dim.x);
 		return HB_MC_FAIL;
 	}
-	
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
-	
-	for (int i = 0; i < num_tiles; i++) { /* initialize tiles */
-		hb_mc_tile_freeze_dep(*fd, tiles[i].x, tiles[i].y);
-		hb_mc_tile_set_group_origin(*fd, tiles[i].x, tiles[i].y, tiles[i].origin_x, tiles[i].origin_y);
+	if (tg->dim.y > device->mesh->dim.y){
+		bsg_pr_err("%s: tile group Y dimension (%d) larger than mesh Y dimension (%d).\n", __func__, tg->dim.y, device->mesh->dim.y);
+		return HB_MC_FAIL;
 	}
+	for (hb_mc_idx_t org_y = device->mesh->origin.y; org_y <= (device->mesh->origin.y + device->mesh->dim.y - tg->dim.y); org_y++){
```

Comment:
What exactly is this looping over? I wan't to say the mesh but I'm honestly not sure.

Reply:
Yes basically these outer loops are iterating over the entire mesh to find the origin of the new tile group we want to allocate, inside there are another two loops that check to see if all the tiles in a tile group starting from this origin are free, I'll add some comments on this, I agree it is confusing.

Reply:
OK cool. Thank you!